tt_metal/impl/buffers/global_semaphore.hpp

```diff
@@ -0,0 +1,42 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <cstdint>
+#include <memory>
+
+#include "tt_metal/common/core_coord.hpp"
+#include "tt_metal/impl/buffers/buffer_constants.hpp"
+#include "tt_metal/llrt/hal.hpp"
+
+namespace tt::tt_metal {
+
+inline namespace v0 {
+class Buffer;
+class Device;
+}  // namespace v0
+
+class GlobalSemaphore {
+   public:
+    GlobalSemaphore(Device *device, CoreRangeSet cores, uint32_t initial_value, BufferType buffer_type = BufferType::L1);
+
+    static std::shared_ptr<GlobalSemaphore> create(
+        Device *device, CoreRangeSet cores, uint32_t initial_value, BufferType buffer_type = BufferType::L1);
+
+    DeviceAddr address() const;
+
+    void initialize();
```

Comment:
Would this ever be expected to be called by a user? Why is it called by the constructor if it's a public method? Can it be called multiple times? I'd suggest either moving this directly into the constructor, or not calling it from the constructor if it's meant to be called separately.

Reply:
I added it as a public method to support the use case of if the user ever needed to re-initialize the semaphore to a known value sometime after the semaphore was created.
This seemed like a potentially useful function to support, though I currently don't know if it will actually be used since in typical operation, a semaphore would be reset back to it's original state by the ops that use it.
I can remove it from being a public api for now until there is an actual need for it?

Reply:
Is there a better name for this than `initialize`? Is this like `acquire`?

Reply:
This will be part of our api going forward correct?  If we need to reuse the global symaphore, maybe the right thing is ResetGlobalSymaphore(GlobalSymaphoreHandle opaque_handle_here)

In general we are doing a functional api for the user that receives handles and operates on those handles and explicitly hides the class from the user.

Reply:
Eyon's naming suggestion makes sense. If we aren't doing the opaque type now, then I can rename the member function to something along the lines of `reset`/`reset_semaphore_value`. This is just writing the original initial value to the sem locations in memory.